api/storage/client.go

```diff
@@ -126,50 +150,248 @@ func UploadArtifactWithOpts(ctx context.Context, c Client, namespace, path strin
+	NewerThan time.Time
+
+	// Verify digest of the cached or downloaded data.
+	ExpectedSHA256 string
+
+	// Receives human-readable debug messages.
+	Logf func(string, ...interface{})
+}
+
+type CacheInfo struct {
+	CachedAt time.Time
+}
+
+// Download the content from an arbitrary URL and cache it for fast access.
+//
+// The content at the URL is assumed to be immutable. If content is already present in the artifact cache for the given URL it will be used instead.
+//
+// Returns CacheSourceError if it fails downloading from sourceURL.
+func CacheURL(ctx context.Context, cli Client, sourceURL string, opts CacheURLOpts) (io.ReadCloser, CacheInfo, error) {
+	logf := opts.Logf
+	if logf == nil {
+		logf = func(string, ...interface{}) {}
+	}
+
+	parsedURL, err := url.Parse(sourceURL)
+	if err != nil {
+		return nil, CacheInfo{}, status.Errorf(codes.InvalidArgument, "invalid URL format: %v", err)
+	}
+
+	labelFilter := []*stdlib.LabelFilterEntry{{Name: cacheSourceURLLabel, Value: sourceURL, Op: stdlib.LabelFilterEntry_EQUAL}}
+	if opts.ExpectedSHA256 != "" {
+		labelFilter = append(labelFilter, &stdlib.LabelFilterEntry{Name: artifactDigestLabel, Value: opts.ExpectedSHA256, Op: stdlib.LabelFilterEntry_EQUAL})
+	}
+
+	listResp, err := cli.Artifacts.ListArtifacts(ctx, &storagev1beta.ListArtifactsRequest{
+		Namespaces:  []string{cacheArtifactNamespace},
+		LabelFilter: labelFilter,
+	})
+	if err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to list artifacts: %w", err)
+	}
+
+	var newest *storagev1beta.Artifact
+	for _, art := range listResp.Artifacts {
+		if newest == nil || art.GetCreatedAt().AsTime().After(newest.GetCreatedAt().AsTime()) {
+			newest = art
+		}
+	}
+
+	if newest != nil {
+		if newest.GetCreatedAt().AsTime().After(opts.NewerThan) {
+			logf("Loading from cache (cached at %v)...\n", newest.GetCreatedAt().AsTime())
+
+			r, ai, err := ResolveArtifactWithOpts(ctx, cli, newest.GetNamespace(), newest.GetPath(), ResolveArtifactOpts{})
+			if err != nil {
+				return nil, CacheInfo{}, err
+			}
+
+			return r, CacheInfo{CachedAt: ai.CreatedAt}, nil
+		} else {
+			logf("Content cached at %v is too old, loading from source...\n", newest.GetCreatedAt().AsTime())
+			// Fallthrough
+		}
+	} else {
+		logf("Artifact not found in cache; loading from source...\n")
+		// Fallthough
+	}
+
+	cachePath := cacheArtifactPath(time.Now(), parsedURL)
+	labels := map[string]string{cacheSourceURLLabel: sourceURL}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", sourceURL, nil)
+	if err != nil {
+		return nil, CacheInfo{}, fmt.Errorf("failed to prepare request: %w", err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, CacheInfo{}, CacheSourceError{fmt.Errorf("failed to send request: %w", err), 0}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, CacheInfo{}, CacheSourceError{fmt.Errorf("remote server returned status code %d", resp.StatusCode), resp.StatusCode}
```

Comment:
Construction feels clunky.
I'd only use a custom error for the case of a bad status code.
Then, we only need to pass the code to a constructor and the message is fixed for the error.

Reply:
Same as above.